File: 0052-n-queens-ii/0052-n-queens-ii.java
class Solution {
  public boolean inBound(int c, int n) {
    return c >= 0 && c < n;
  }
  
  public void dfs(List<Integer> seq, List<List<Integer>> seqs, int n) {
    if (seq.size() == n) {
      seqs.add(seq);
      return;
    }
    Set redSet = new HashSet<Integer>(seq);
    for (int r = 0; r < seq.size(); r++) {
      int c = seq.get(r);
      if (inBound(c + seq.size() - r, n)) redSet.add(c + seq.size() - r);
      if (inBound(c - seq.size() + r, n)) redSet.add(c - seq.size() + r);
    }
    for (int i = 0; i < n; i++) {
      List newSeq = new ArrayList<Integer>();
      newSeq.addAll(seq);
      newSeq.add(i);
      if (!redSet.contains(i)) dfs(newSeq, seqs, n);
    }
  }
  
  public int totalNQueens(int n) {
    List seqs = new ArrayList();
    for (int i = 0; i < n; i++) {
      List starter = Arrays.asList(i);
      dfs(starter, seqs, n);
    }
    return seqs.size();
  }
}

File: 0052-n-queens-ii/0052-n-queens-ii.go
func totalNQueens(n int) int {
  ans := 0
  for i:=0;i<n;i++ {
    dfs([]int{i}, n, &ans)
  }
  return ans
}

func dfs(seq []int, n int, ans *int)  {
  if len(seq) == n {
    *ans++
    return
  }
  redSet := map[int]bool{}
  for _, s := range seq {
    redSet[s] = true
  }
  for r:=0;r<len(seq);r++ {
    if inBound(seq[r]+len(seq)-r, n) {
      redSet[seq[r]+len(seq)-r] = true
    }
    if inBound(seq[r]-len(seq)+r, n) {
      redSet[seq[r]-len(seq)+r] = true
    }
  }
  for i:=0;i<n;i++ {
    newSeq := append(seq, i)
    _, ok := redSet[i]
    if !ok {
      dfs(newSeq, n, ans)
    }
  }
  return
}

func inBound(c int, n int) bool {
  return c>=0 && c<n
}
